REL-2852 improve mean parallactic angle calculation usability

tpolecat opened the pull request:
This PR improves the usability for mean parallactic angle computations as follows:

- The duration used for the computation is retained, even if it is a snapshot of the current remaining time. We do this by storing it as a negative value, in order to maintain binary compatibility with the existing `SchedulingBlock` class.
- When no duration is specified, remaining observe time is used instead and is *not* recorded in the scheduling block. This is a corner case that happens only once in practice, when you first select the mean parallactic angle option. Since this is never the final answer I don't think it matters … it would end up being a potentially risky change to update the scheduling block in this case.

This needs monkey testing.

swalker2m commented:
Technically we maintain binary compatibility but what happens when the previous OT version edits an observation with a negative scheduling duration?

In the end, another patch release feature will require breaking binary compatibility anyway if we want to take the opportunity to do this in a more explicit way.

cquiroz commented:
code-wise looks good, I can't comment on the details of the changes

tpolecat commented:
Ah, good point about the old OT. That's not going to work.

tpolecat commented:
Ok since this is going to be a breaking release anyway I can re-do this in a better way, but not in time for Monday's test release because it's going to require writing a new data migration. Not sure which way to go.

swalker2m commented:
My vote would be to hold up the test release until this is ready.  We'll have to do it differently for the `develop` branch anyway so we might as well just merge the same thing to both branches.  Perhaps if it were ready except for the data migration, I could work on that Monday morning and we might be able to do a test release regardless.  If not, then Tuesday.

What do you think @andrewwstephens / @arturog8m ?

tpolecat commented:
That sounds reasonable. I'll see how much I can get done this afternoon.

tpolecat commented:
Ok @swalker2m I changed the model such that `SchedulingBlock.duration` is an ADT with three cases. PIO serialization works but there is no handler in `To2016B` and the existing scheduling block migration fails at the moment. I will look at these Monday morning, or you can.

tpolecat commented:
@swalker2m hey is `SPProgram.VERSION` going to change to be `2016B-2`? I need to distinguish pre and post-patch XML files in the updater.

tpolecat commented:
Ok updated as follows:
- We now have an ADT `SchedulingBlock.Duration` that distinguishes `Unstated`, `Implicit`, and `Explicit` durations rather than encoding this via negative numbers.
- Changed `SPProgram.VERSION ` to `2016B-2` and added a new migration.

swalker2m commented:
Looks great. 👍

cquiroz commented:
Would you also integrate this PR into develop?
